File: 347-top-k-frequent-elements/top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
      const intCounts = new Object();
  for (let i = 0; i < nums.length; i++) {
    if (intCounts[nums[i]]) {
      intCounts[nums[i]]++;
    } else {
      intCounts[nums[i]] = 1;
    }
  }

  const frequency = new Map();
  for (let i = 1; i < nums.length + 1; i++) {
    frequency.set(i, []);
  }

  for (let i = 0; i < Object.values(intCounts).length; i++) {
    if (frequency.has(Object.values(intCounts)[i])) {
      frequency
        .get(Object.values(intCounts)[i])
        .push(Object.keys(intCounts)[i]);
    }
  }
  const simplifiedFreq = new Map();
  for (let [key, value] of frequency) {
    if (value.length > 0) {
      simplifiedFreq.set(key, value);
    }
  }
  console.log(simplifiedFreq);
  const arr = Array.from(simplifiedFreq.values()).flat();
  const output = [];
  for (let i = 0; i < k; i++) {
    output.push(parseInt(arr.pop()));
  }
  return output;
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
 public int[] topKFrequent(int[] nums, int k) {
    int[] output = new int[k];
    HashMap<Integer, Integer> numbers = new HashMap<Integer, Integer>();
    for (int i = 0; i < nums.length; i++) {
      numbers.put(nums[i], numbers.getOrDefault(nums[i], 0) + 1);
    }
    List<List<Integer>> bucket = new ArrayList<>();
    for (int i = 0; i < nums.length + 1; i++) {
      // nums.length + 1 because if nums = [1, 1, 1, 1, 1], frequency of 1s will be 5
      bucket.add(new ArrayList<>());
    }
    for (int key : numbers.keySet()) {
      bucket.get(numbers.get(key)).add(key);
    }
    int idx = 0;
    for (int i = bucket.size() - 1; i >= 0; i--) {
      if (bucket.get(i).size() > 0) {
        for (int num : bucket.get(i)) {
          output[idx] = num;
          idx++;
          if (idx == k) {
            break;
          }
        }
        if (idx == k) {
          break;
        }
      }
    }
    return output;
  }
}